lms/djangoapps/completion/models.py

```diff
@@ -0,0 +1,120 @@
+            * user (django.contrib.auth.models.User): The user for whom the
+              completion is being submitted.
+            * course_key (opaque_keys.edx.keys.CourseKey): The course in
+              which the submitted block is found.
+            * block_key (opaque_keys.edx.keys.UsageKey): The block that has had
+              its completion changed.
+            * completed (float in range [0.0, 1.0]): The fractional completion
+              value of the block (0.0 = incomplete, 1.0 = complete).
+
+        Return Value:
+            (BlockCompletion, bool): A tuple comprising the created or updated
+            BlockCompletion object and a boolean value indicating whether the value
+
+        Raises:
+
+            ValueError:
+                If the wrong type is passed for one of the parameters.
+
+            django.core.exceptions.ValidationError:
+                If a float is passed that is not between 0.0 and 1.0.
+
+            django.db.DatabaseError:
+                If there was a problem getting, creating, or updating the
+                BlockCompletion record in the database.
+
+                This will also be a more specific error, as described here:
+                https://docs.djangoproject.com/en/1.11/ref/exceptions/#database-exceptions.
+                IntegrityError and OperationalError are relatively common
+                subclasses.
+        """
+
+        # Raise ValueError to match normal django semantics for wrong type of field.
+        if not isinstance(course_key, CourseKey):
+            raise ValueError(
+                "course_key must be an instance of `opaque_keys.edx.keys.CourseKey`.  Got {}".format(type(course_key))
+            )
+        try:
+            block_type = block_key.block_type
+        except AttributeError:
+            raise ValueError(
+                "block_key must be an instance of `opaque_keys.edx.keys.UsageKey`.  Got {}".format(type(block_key))
+            )
+
+        obj, isnew = self.get_or_create(
+            user=user,
+            course_key=course_key,
+            block_type=block_type,
+            block_key=block_key,
+            defaults={'completed': completed},
+        )
+        if not isnew and obj.completed != completed:
+            obj.completed = completed
+            obj.full_clean()
+            obj.save()
+        return obj, isnew
+
+
+class BlockCompletion(TimeStampedModel, models.Model):
+    """
+    Track completion of completable blocks.
+    """
+    user = models.ForeignKey(User)
+    course_key = CourseKeyField(max_length=255)
+    block_key = UsageKeyField(max_length=255)
+    block_type = models.CharField(max_length=64)
+    completed = models.FloatField(validators=[validate_ratio])
+
+    objects = BlockCompletionManager()
+
+    class Meta(object):
+        index_together = [
+            ('course_key', 'block_type', 'user'),
+        ]
+
+        unique_together = [
+            ('user', 'course_key', 'block_key')
+        ]
```

Comment:
This will work as an index on:

* all completions for a given user
* all completions in a course for a given user
* the specific completion for a particular block for a particular user.

Things that aren't indexed:

* All completions for a particular block across all users
* All completions modified since a given time at any granularity.  For aggregating smartly, do we need to add a `(user, course, timestamp)` key?  What about `(course, timestamp)`?  I think we might need the former, but not the latter, because I think we will always need to compare the timestamp against the specific user's aggregations to know whether to update them.

Reply:
This might be a naive question, but are block key's globally unique, and would one need to know to use the course key to make use of this index for the block key?

Reply:
I believe block keys are globally unique for split mongo, but they drop the run value from the course key in old-mongo, so if the same block exists in two runs of the same course (highly likely), then they need to be disambiguated.

Reply:
Update to include an index for recently modified values and to rename the validator.